hypothesis-python/tests/cover/test_database_backend.py

```diff
@@ -160,3 +168,128 @@ def test_multiplexed_dbs_read_and_write_all():
     multi.delete(b"c", b"cc")
     for db in (a, b, multi):
         assert set(db.fetch(b"c")) == set()
+
+
+def test_ga_require_readonly_wrapping():
+    database = GitHubArtifactDatabase("test", "test")
+    # save, move and delete can only be called when wrapped around ReadonlyDatabase
+    with pytest.raises(RuntimeError):
+        database.save(b"foo", b"bar")
+    with pytest.raises(RuntimeError):
+        database.move(b"foo", b"bar", b"foobar")
+    with pytest.raises(RuntimeError):
+        database.delete(b"foo", b"bar")
+
+    # check that the database silently ignores writes when wrapped around ReadOnlyDatabase
+    database = ReadOnlyDatabase(database)
+    database.save(b"foo", b"bar")
+    database.move(b"foo", b"bar", b"foobar")
+    database.delete(b"foo", b"bar")
+
+
+def ga_mock_empty_artifact() -> Path:
+    temp_dir = tempfile.mkdtemp()
+    path = Path(temp_dir) / "github-artifacts"
+    path.mkdir(parents=True, exist_ok=True)
+    zip_path = path / f"{datetime.datetime.now().isoformat()}.zip"
+
+    with zipfile.ZipFile(zip_path, "w"):
+        pass
+
+    return path
+
+
+def test_ga_empty_read():
+    path = ga_mock_empty_artifact()
+    database = GitHubArtifactDatabase("test", "test", path=path)
+    assert list(database.fetch(b"foo")) == []
+
+
+def test_ga_initialize():
+    path = ga_mock_empty_artifact()
+    database = GitHubArtifactDatabase("test", "test", path=path)
+    # Trigger initialization
+    database.fetch(b"")
+    root1 = database._artifact
+    # Should not trigger initialization
+    database.fetch(b"")
+    root2 = database._artifact
+    assert root1 == root2
+
+
+def test_ga_no_artifact():
+    tmp_dir = Path(tempfile.mkdtemp())
+    database = GitHubArtifactDatabase("test", "test", path=tmp_dir)
+    # Check that the database raises a warning
+    with pytest.warns(HypothesisWarning):
+        assert list(database.fetch(b"")) == []
+    assert database._disabled is True
+
+
+class GitHubArtifactMocks(RuleBasedStateMachine):
+    """
+    Setups a mock GithubArtifactDatabase from an artifact
+    generated from a DirectoryBasedExampleDatabase.
+    """
+    def __init__(self):
+        super().__init__()
+        self.temp_directory = Path(tempfile.mkdtemp())
+        self.path = self.temp_directory / "github-artifacts"
+
+        # This is where we will store the contents for the zip file
+        self.zip_destination = self.path / f"{datetime.datetime.now().isoformat()}.zip"
+
+        # And this is where we want to create it
+        self.zip_content_path = self.path / datetime.datetime.now().isoformat()
+        self.zip_content_path.mkdir(parents=True, exist_ok=True)
+
+        # We use a DirectoryBasedExampleDatabase to create the contents
+        self.directory_db = DirectoryBasedExampleDatabase(str(self.zip_content_path))
```

Comment:
We usually prefer an `InMemoryExampleDatabase` for testing, since it's a little cheaper and faster.  Any reason not to do that here?

Reply:
It's possible, but I use `DirectoryBasedExampleDatabase` because I can just archive it and pass that as the artifact.

Reply:
Oh, nice.  Comment to that effect then?